### ISSUE TITLE
Severe Perf issue in Statcast

### ISSUE THREAD

**Issue by markdelgrande** [jldbc/pybaseball#154]
When calling data = statcast('2020-08-01', '2020-10-05') on a large query it just runs forever now.

**Comment by jldbc** [comment]
What version are you using @markdelgrande? I'm wondering if this is coming from #146 or #152, I couldn't reproduce before upgrading to the most recent release

**Comment by markdelgrande** [comment]
nah I just reverted to version 2.0.0 and it works now, its broken with the new update

**Comment by jldbc** [comment]
Mine actually completed after a while. Maybe the verbose flag is broken and it's just not giving progress updates as it works through the dates? Or does yours never finish

**Comment by markdelgrande** [comment]
Mine never finishes I let it run for like 30 mins and nothing happened ran the same code but with the old stuff and its all good now.

**Comment by markdelgrande** [comment]
Im fine running the old stuff, but I thought id just let you guys know in case you wanted to fix it.

**Comment by jldbc** [comment]
Sounds good, thanks for flagging. @schorrm we should still look into this, I can't exactly reproduce this but it does at least seem to be hanging/slowing down for me and not printing status updates

**Comment by schorrm** [comment]
This is very, very wrong -- I have similar behavior. I don't have anything concrete to suggest, I may consider nuking this and restoring to the old one if I don't get any ideas. @TheCleric I think you were the last one to touch?

**Comment by jldbc** [comment]
I support reverting this and taking a slower look at the refactoring changes if there are no theories for a quick fix

**Comment by bdilday** [comment]
I think it's here where it's doing a loop over the rows of a large dataframe to auto-detect data type https://github.com/jldbc/pybaseball/blob/master/pybaseball/datahelpers/postprocessing.py#L26

I did some profiling with `cProfile` which suggests this and I think logically this makes sense as a bottleneck

**Comment by schorrm** [comment]
Makes sense. Can we address this with convert_dtypes? Is that still too slow?

**Comment by schorrm** [comment]
Nuking the post-processing fails to fix this.

**Comment by schorrm** [comment]
I'll do a hotfix which is just going to roll back the entire file.

**Comment by TheCleric** [comment]
I would say that reverting the whole file is a bit extreme. Why don't we try to figure out the actual problem?

**Comment by schorrm** [comment]
We can, but I'm fixing the bug in the release and then once the issue is fixed we'll do another update.

**Comment by TheCleric** [comment]
So after some testing on my machine, here's what I'm seeing. I ran the following command to test:


```python
import timeit

timeit.timeit("data = statcast('2020-08-01', '2020-10-05')", "from pybaseball import statcast", number=1)
```

v2.1 on my machine takes 240 seconds.
v2.0 on my machine takes 99 seconds.
So we obviously regressed in terms of performance.

As well I inadvertently broke the print statements that communicated progress, so this likely feels like it takes even longer than it is.

Interestingly in my experiment, I changed the days that divide a small_request from a large_request from 5 to 365, and after that, I got v2.1 down to 47 seconds.

So for a quick fix we could just make that dividing line much larger. I'm going to keep digging, but if I had to guess it's the fact that we're running the postprocessing per small request, and not simply at the end.

**Comment by TheCleric** [comment]
Also @schorrm I appreciate wanting to get a quick fix out, but when we do pr the fix it will now be very difficult to see what was actually fixed.

**Comment by TheCleric** [comment]
Yeah it's definitely the postprocessor. Dropping it and replacing it with a general pd.to_numeric coercion gets it down to 88 seconds. Will look at what we can do to speed it up, or I'll submit a PR removing it.

**Comment by jldbc** [comment]
re: the number of days large requests get chunked into, 5 isn't completely arbitrary. Savant returns incomplete results if you ask for too many days of data. Requests need to be small enough where you can essentially guarantee you wont be quietly losing data. In 2017 that number was about 5 days of data, I haven't checked whether that's changed since though.

**Comment by TheCleric** [comment]
> re: the number of days large requests get chunked into, 5 isn't completely arbitrary. Savant returns incomplete results if you ask for too many days of data. Requests need to be small enough where you can essentially guarantee you wont be quietly losing data. In 2017 that number was about 5 days of data, I haven't checked whether that's changed since though.

Well that may be why it comes back faster than, if it just drops data. 😆

**Comment by schorrm** [comment]
Should we be multiprocessing this, perhaps?

**Comment by bdilday** [comment]
> Nuking the post-processing fails to fix this.

@schorrm what you do to test that and what performance did you see? It seemed to fix the issue for me. 

also, I think we should add an integration test for this

**Comment by schorrm** [comment]
@bdilday using the `convert_dtypes()` at the end worked?

**Comment by schorrm** [comment]
And did the print issues go away too?

**Comment by bdilday** [comment]
> @bdilday using the `convert_dtypes()` at the end worked?

oh no, my mistake I didn't try that I just removed call to `try_parse`, i.e., 

```
def get_statcast_data_from_csv(
        csv_content: str,
        null_replacement: Union[str, int, float, datetime] = np.nan,
        known_percentages: List[str] = []
    ) -> pd.DataFrame:
    data = pd.read_csv(io.StringIO(csv_content))
    return postprocessing.try_parse_dataframe(
        data,
        null_replacement=null_replacement,
        known_percentages=known_percentages
    )
```

becomes, 

```
def get_statcast_data_from_csv(
        csv_content: str,
        null_replacement: Union[str, int, float, datetime] = np.nan,
        known_percentages: List[str] = []
    ) -> pd.DataFrame:
    data = pd.read_csv(io.StringIO(csv_content))
    return data
```

and that seemed to address the timing issue

**Comment by TheCleric** [comment]
> also, I think we should add an integration test for this

We do have some integration tests around this, but they are purposefully smaller so as not to take up too much time per @schorrm

**Comment by TheCleric** [comment]
I think I am VERY close to a solution that is much faster. Just working through some edge cases.

**Comment by TheCleric** [comment]
> re: the number of days large requests get chunked into, 5 isn't completely arbitrary. Savant returns incomplete results if you ask for too many days of data. Requests need to be small enough where you can essentially guarantee you wont be quietly losing data. In 2017 that number was about 5 days of data, I haven't checked whether that's changed since though.

@jldbc it looks like Statcast returns a max of 40000 rows, which explains the missing data (and relative speed!) when trying to get too much at once. In experimenting though it looks like you can fit 8 days of data into that window (at least going by the entire 2019 season). So I'll up it to 8 to help with performance.

**Comment by jldbc** [comment]
I'd want to see how much wiggle room that has over a larger time period (including the current season with all its extra double headers) to be extremely confident. I'd rather be slow than return bad data

**Comment by schorrm** [comment]
@TheCleric - what about 2020? More frequent games etc. If it works for 2020, then good with me if @jldbc signs off on this, just please put this comment of yours as a comment there.
(Checking if `len(df) == 40_000` could be an option?)

**Comment by schorrm** [comment]
@jldbc - I wrote this at the same time as yours, I commented and then saw your comment... Great minds think alike I guess
I am concerned a bit less about the raw time and more about a smooth behavior

**Comment by TheCleric** [comment]
> @TheCleric - what about 2020? More frequent games etc. If it works for 2020, then good with me if @jldbc signs off on this, just please put this comment of yours as a comment there.

Good call. Testing with 2020 yields good results as well. Most returned in a single 8 day period was @ 38000.

> (Checking if `len(df) == 40_000` could be an option?)

I thought about that as well. Since some of the 2020 data gets close to this, I went ahead and added code that detects this and breaks it down.

**Comment by TheCleric** [comment]
Good news. I've got a new solution that gets the same parsing we were getting out of the old postprocessor, but in much less time. It's still a LITTLE slower than the 2.0 code, but that's expected as the 2.0 code wasn't as accurate in data type parsing as 2.1 was (hence why it was much slower).

On my machine it takes about 120 seconds to do the timeit query above based on the original bug reported. This is about 20% slower than 2.0, but is still half the time that 2.1 was taking, so I count that as a win.

As well, I think I can retrofit some of the tweaks I figured out into our other parsing code (such as FanGraphs) to speed that up as well. I'll get this fix in as a PR as soon as I'm 100% confident that I didn't accidentally break anything else.

After that I'll submit a separate PR to retrofit some old code with this faster parsing.
